geo/src/algorithm/voronoi_diagram.rs

```diff
@@ -0,0 +1,970 @@
+/// and thus the Voronoi Diagram can be created from the Delaunay Triangulation.
+pub fn compute_voronoi_components_from_delaunay<T: GeoFloat>(
+    triangles: &[Triangle<T>],
+    clipping_mask: Option<&Polygon<T>>,
+) -> Result<VoronoiComponents<T>>
+where
+    f64: From<T>,
+{
+    let delaunay_triangles: Vec<DelaunayTriangle<T>> =
+        triangles.iter().map(|x| (*x).into()).collect();
+
+    if delaunay_triangles.is_empty() {
+        return Ok(VoronoiComponents {
+            delaunay_triangles: vec![],
+            vertices: vec![],
+            lines: vec![],
+            neighbours: vec![],
+        });
+    }
+
+    let clipping_mask = match clipping_mask {
+        Some(mask) => mask.clone(),
+        None => create_clipping_mask(triangles)?,
+    };
+
+    // The centres of the delaunay circumcircles form the vertices of the
+    // voronoi diagram
+    let mut vertices: Vec<Coord<T>> = Vec::new();
+    for tri in delaunay_triangles.iter() {
+        vertices.push(
+            tri.get_circumcircle_center()
+                .map_err(VoronoiDiagramError::DelaunayError)?,
+        );
+    }
+
+    // Find the shared edges
+    let mut shared = find_shared_edges(&delaunay_triangles);
+
+    // Create the lines joining the vertices
+    let mut voronoi_lines: Vec<Line<T>> = Vec::new();
+    for neighbour in shared.neighbours.iter() {
+        if let (Some(first_vertex), Some(second_vertex)) = (neighbour.0, neighbour.1) {
+            voronoi_lines.push(Line::new(vertices[first_vertex], vertices[second_vertex]));
+        }
+    }
+
+    voronoi_lines.extend(construct_edges_to_inf(
+        &delaunay_triangles,
+        &vertices,
+        &mut shared.shared_edges,
+        &shared.neighbours,
+        &clipping_mask,
+    )?);
+
+    Ok(VoronoiComponents {
+        delaunay_triangles: triangles.to_vec(),
+        vertices,
+        lines: voronoi_lines,
+        neighbours: shared.neighbours,
+    })
+}
+
+fn create_clipping_mask<T: GeoFloat>(triangles: &[Triangle<T>]) -> Result<Polygon<T>>
+where
+    f64: From<T>,
+{
+    let mut pts: Vec<Point<T>> = Vec::new();
+    for tri in triangles.iter() {
+        pts.extend(tri.to_array().iter().map(|x| Point::from(*x)));
+    }
+    let bounds = MultiPoint::new(pts)
+        .bounding_rect()
+        .ok_or(VoronoiDiagramError::CannotDetermineBoundsFromClipppingMask)?;
+
+    let expand_factor = T::from(DEFAULT_SUPER_TRIANGLE_EXPANSION)
+        .ok_or(VoronoiDiagramError::CannotConvertBetweenGeoGenerics)?;
+
+    let width = bounds.width() * expand_factor;
+    let height = bounds.height() * expand_factor;
+    let bounds_min = bounds.min();
+
+    Ok(Polygon::from(Rect::new(
+        coord! { x: bounds_min.x - width, y: bounds_min.y - height},
+        coord! { x: bounds_min.x + width, y: bounds_min.y + height},
+    )))
+}
+
+#[derive(Debug, Clone)]
+struct Neighbours<T: GeoFloat> {
+    pub neighbours: Vec<(Option<usize>, Option<usize>)>,
```

Comment:
pub members of a non-pub struct?

Reply:
@jayvdb  I'm interested in your perspective on this one.  Originally the `find_shared_edges` function returned `(Vec<(Option<usize>, Option<usize>)>, Vec<Line<T>>`.  Clippy suggested to move it to a struct which is why `Neighbours` is there, also I should probably rename to american english.  I did not intend on `Neighbours` being public, but also dont see the benefit in implementing get/set methods.  This is why there is public members of a non public struct.

Reply:
as currently written, the `pub` on the members has no effect - they can be removed, and the members can still be accessed within the module.

Rename `Neighbours` to be `SharedEdgesData` due to it being returned from fn `find_shared_edges` ?

And maybe add `type Neighbour = (Option<usize>, Option<usize>); type Neighbours = Vec<Neighbour>;` ?